bin/connect-pair.py

```diff
@@ -1,21 +1,38 @@
 #!/usr/bin/python
 
-# Connects two copy-paste samples running on localhost.
+# Connects two SOCKS adventure instances running on localhost.
 # TODO: add host/port args.
 
+import select
 import socket
-import time
 
-first = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
-first.connect(("localhost", 9000))
+getter = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+getter.connect_ex(('localhost', 9000))
+getter.setblocking(False)
 
-second = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
-second.connect(("localhost", 9010))
+giver = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+giver.connect(('localhost', 9010))
+giver.setblocking(False)
 
-first.send("GET\n")
-time.sleep(0.5)
-first_sdp = first.recv(4096)
-second.send("GIVE " + first_sdp + "\n")
-time.sleep(0.5)
-second_sdp = second.recv(4096)
-first.send(second_sdp)
+getter.sendall('get\n')
+giver.sendall('give\n')
+
+getterOpen = True
+giverOpen = True
+while getterOpen and giverOpen:
+  inready, outready, exceptready = select.select([getter, giver], [], [])
+  for s in inready:
+    raw = s.recv(4096)
```

Comment:
Perhaps make the sockets nonblocking (https://docs.python.org/2/howto/sockets.html#non-blocking-sockets) and `read()` until you don't get data back (or just < 4k).  Otherwise you may splice the last line of a large block with a newline below.